micrometer-core/src/main/java/io/micrometer/core/instrument/binder/system/MemoryMetrics.java

```diff
@@ -0,0 +1,174 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.system;
+
+import io.micrometer.common.lang.NonNullApi;
+import io.micrometer.common.lang.NonNullFields;
+import io.micrometer.common.lang.Nullable;
+import io.micrometer.core.instrument.*;
+import io.micrometer.core.instrument.binder.BaseUnits;
+import io.micrometer.core.instrument.binder.MeterBinder;
+
+import java.lang.management.*;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.List;
+
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Record metrics that report memory usage.
+ *
+ * @author Fábio C. Martins
+ */
+@NonNullApi
+@NonNullFields
+public class MemoryMetrics implements MeterBinder {
+
+    /**
+     * List of public, exported interface class names from supported JVM implementations.
+     */
+    private static final List<String> OPERATING_SYSTEM_BEAN_CLASS_NAMES = Arrays.asList(
+            "com.ibm.lang.management.OperatingSystemMXBean", // J9
+            "com.sun.management.OperatingSystemMXBean" // HotSpot
+    );
+
+    private final OperatingSystemMXBean operatingSystemBean;
+
+    @Nullable
+    private final Class<?> operatingSystemBeanClass;
+
+    private final Iterable<Tag> tags;
+
+    @Nullable
+    private final Method committedVirtualMemorySize;
+
+    @Nullable
+    private final Method totalSwapSpaceSize;
+
+    @Nullable
+    private final Method freeSwapSpaceSize;
+
+    @Nullable
+    private final Method freePhysicalMemorySize;
+
+    @Nullable
+    private final Method totalPhysicalMemorySize;
+
+    public MemoryMetrics() {
+        this(emptyList());
+    }
+
+    public MemoryMetrics(Iterable<Tag> tags) {
+        this.tags = tags;
+        this.operatingSystemBean = ManagementFactory.getOperatingSystemMXBean();
+        this.operatingSystemBeanClass = getFirstClassFound(OPERATING_SYSTEM_BEAN_CLASS_NAMES);
+        this.committedVirtualMemorySize = detectMethod("getCommittedVirtualMemorySize");
+        this.totalSwapSpaceSize = detectMethod("getTotalSwapSpaceSize");
+        this.freeSwapSpaceSize = detectMethod("getFreeSwapSpaceSize");
+        this.freePhysicalMemorySize = detectMethod("getFreePhysicalMemorySize");
+        this.totalPhysicalMemorySize = detectMethod("getTotalPhysicalMemorySize");
+    }
+
+    @Override
+    public void bindTo(MeterRegistry registry) {
+        Runtime runtime = Runtime.getRuntime();
+        if (committedVirtualMemorySize != null) {
+            Gauge.builder("system.virtualmemory.commited", operatingSystemBean, x -> invoke(committedVirtualMemorySize))
```

Comment:
Maybe split the words virtual and memory using `virtual_memory` for better readability?

Reply:
And I also think that this should be `process.virtual_memory.committed`. The javadoc on `getCommittedVirtualMemorySize` says `Returns the amount of virtual memory that is guaranteed to be available to the running process.`. So, maybe `process` namespace suits it better than `system.`

Reply:
I renamed the gauge.
I also noticed that other metrics are referenced by the class _PrometheusSample_ should I add _MemoryMetrics_ to it? What is its purpose?